plan/preprocess.go

```diff
@@ -18,12 +18,98 @@ import (
 	"github.com/pingcap/tidb/ast"
 	"github.com/pingcap/tidb/context"
 	"github.com/pingcap/tidb/infoschema"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/privilege"
 )
 
 // Preprocess does preprocess work for optimizer.
 func Preprocess(node ast.Node, info infoschema.InfoSchema, ctx context.Context) error {
 	if err := ResolveName(node, info, ctx); err != nil {
 		return errors.Trace(err)
 	}
+
+	if checker := privilege.GetPrivilegeChecker(ctx); checker != nil {
+		if err := CheckPrivilege(node, ctx, checker); err != nil {
+			return errors.Trace(err)
+		}
+	}
 	return nil
 }
+
+// CheckPrivilege is exposed for testing.
+func CheckPrivilege(node ast.Node, ctx context.Context, checker privilege.Checker) error {
+	privChecker := privilegeChecker{
+		Checker: checker,
+		ctx:     ctx,
+		pass:    true,
+	}
+	node.Accept(&privChecker)
+	if !privChecker.pass {
+		return errors.New("check privilege failed!")
+	}
+
+	return nil
+}
+
+// privilegeChecker wraps privilege.Checker and implements ast.Visitor interface.
+type privilegeChecker struct {
+	ctx context.Context
+	privilege.Checker
+	pass bool
+}
+
+func (pc *privilegeChecker) check(t *ast.TableName, priv mysql.PrivilegeType) {
+	// TODO: Call Checker's check function
+}
+
+func (pc *privilegeChecker) Enter(in ast.Node) (ast.Node, bool) {
+	if !pc.pass {
+		return in, true
+	}
+	switch v := in.(type) {
+	case *ast.AdminStmt:
+	case *ast.AlterTableStmt:
+		pc.check(v.Table, mysql.AlterPriv)
+	case *ast.AnalyzeTableStmt:
+		for _, t := range v.TableNames {
+			pc.check(t, mysql.SelectPriv)
+		}
+	case *ast.CreateIndexStmt:
+		pc.check(v.Table, mysql.IndexPriv)
+	case *ast.CreateTableStmt:
+		pc.check(v.Table, mysql.CreatePriv)
+	case *ast.DeleteStmt:
+		for _, t := range v.Tables.Tables {
+			pc.check(t, mysql.DeletePriv)
+		}
+	case *ast.DeleteTableList:
+		for _, t := range v.Tables {
+			pc.check(t, mysql.DeletePriv)
+		}
+	case *ast.DropTableStmt:
+		for _, t := range v.Tables {
+			pc.check(t, mysql.DropPriv)
+		}
+	case *ast.DropIndexStmt:
+		pc.check(v.Table, mysql.IndexPriv)
+	case *ast.FieldList:
+	case *ast.InsertStmt:
+	case *ast.LoadDataStmt:
+	case *ast.SelectStmt:
+		for _, rf := range v.GetResultFields() {
```

Comment:
Not all used table will appear in select stmt, Like `select a.* from a join b`, You can checkout it in `buildDataSource` in logical plan builder.